### PR TITLE
deny both warnings and missing docs

### PR DESCRIPTION
**Pull request by papertigers**
Add some docs for `cargo doc --no-deps` and therefor a meaningful docs.rs page.

Also add some doc tests that show examples as well as verify things are working to some extent.

```
root - rustdev ~/src/rust-zonename (git:master) # cargo test
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running target/debug/deps/zonename-1f1a732f627e4cbc
running 0 tests
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
     Running target/debug/deps/zonename-ae8c98dd60a93b09
running 0 tests
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
   Doc-tests zonename
running 4 tests
test src/lib.rs - getzoneid (line 24) ... ok
test src/lib.rs - getzoneidbyname (line 39) ... ok
test src/lib.rs - getzonenamebyid (line 58) ... ok
test src/lib.rs - getzonename (line 81) ... ok
test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -1,64 +0,0 @@
-[[package]]
```
Comment (bahamas10) [inline Cargo.lock:1]: should this file not be checked into git?

File: README.md [hunk 2]
```diff
@@ -1,8 +1,10 @@
 rust-zonename
 =============
 
-Rust bindings to `getzoneid(3C)`, `getzoneidbyname(3C)`, and
-`getzonenamebyid(3C)` for Illumos based systems.
+Rust bindings to [getzoneid(3C)](https://illumos.org/man/3C/getzoneid),
+[getzoneidbyname(3C)](https://illumos.org/man/3C/getzoneid), and
```
Comment (bahamas10) [inline README.md]: nice to make these links.... i realize the manpage is the same but they should probably point to their respective pages... ie

- https://illumos.org/man/3C/getzoneid
- https://illumos.org/man/3C/getzoneidbyname
- https://illumos.org/man/3C/getzonenamebyid

File: Cargo.toml [hunk 3]
```diff
@@ -5,4 +5,3 @@ authors = ["Dave Eddy <dave@daveeddy.com>"]
 
 [dependencies]
 libc = "0.2.43"
-errno = "0.2.4"
```
Comment (bahamas10) [inline Cargo.toml:8]: nice! didn't know this was builtin this is so much nicer

File: README.md [hunk 4]
```diff
@@ -1,8 +1,10 @@
...
+[getzonenamebyid(3C)](https://illumos.org/man/3C/getzoneid) for illumos based
+systems.
```
Comment (bahamas10) [inline README.md]: at the bottom of the file the function signatures and `errno` crate note should probably be removed

File: src/lib.rs [hunk 5]
```diff
@@ -1,59 +1,94 @@
...
 
-pub fn getzoneid() -> Result<i32, Errno> {
-    let zoneid = unsafe {
-        ffi::getzoneid()
-    };
+/// Returns the zone ID of the calling process.
+///
+/// # Example
+/// ```
+/// let zoneid = zonename::getzoneid().expect("failed to get zoneid");
```
Comment (bahamas10) [inline src/lib.rs:24]: this is so cool.  i didn't realize the documentation would also be used for testing that's insanely awesome.

File: src/lib.rs [hunk 6]
```diff
@@ -1,59 +1,94 @@
...
+/// ```
+pub fn getzoneid() -> io::Result<i32> {
```
Comment (bahamas10) [inline src/lib.rs:26]: You don't have to specify the error type when returning an `io::Result`? this is really convenient

File: src/lib.rs [hunk 7]
```diff
@@ -1,59 +1,94 @@
...
+    let len = unsafe { ffi::getzonenamebyid(id, ptr, buf.capacity() as size_t) };
     if len < 0 {
-        return Err(errno());
+        return Err(io::Error::last_os_error());
     }
 
     assert!(len >= 0);
 
+    // Attempt to force unwrap cstring because the kernel really shouldn't be giving us back
+    // invalid utf8 characters
```
Comment (bahamas10) [inline src/lib.rs:72]: +1 to this comment

File: Cargo.lock (hunk 1 above)
Comment (papertigers) [reply Cargo.lock:1]: Libraries don't need Cargo.lock while executables do.
https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html

File: README.md (hunk 2 above)
Comment (papertigers) [reply README.md]: Will fix

File: README.md (hunk 4 above)
Comment (papertigers) [reply README.md]: Will fix

File: src/lib.rs (hunk 5 above)
Comment (papertigers) [reply src/lib.rs:24]: Yeah, these are called "doc tests".  Useful for making sure the examples you provide actually do what you think they do.

File: src/lib.rs (hunk 6 above)
Comment (papertigers) [reply src/lib.rs:26]: Yeah its basically a typedef:
https://doc.rust-lang.org/std/io/type.Result.html

**Comment by papertigers** [comment]
Let me know if you want me to squash my commits.  (I think there is a UI option to actually do that automatically when you merge now)

**Pull request merged**